### DISCUSSION TITLE
ui_xcontrol

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#67] [Announcements]
# ui_xcontrol

开源共享版本说明

https://lv.qizhen.xyz/ui_xcontrol

**Comment by tetsuoyyx** [comment]
阮老师，Xcontrol控件在同一个VI里面可以多次重复使用吗（像自带的那些控件一样，可以放置任意多个）？我试了一下，好像有问题。

**Reply by ruanqizhen** [reply]
Xcontrol 控件可以在同一个VI里面多次重复使用。你具体遇到了什么问题？

**Reply by tetsuoyyx** [reply]
主要是性能问题，用了这个功能以后前面板动作响应变慢慢，有时候会出现个别Xcontrol控件完全不响应的问题。

**Reply by ruanqizhen** [reply]
听起来像是 XControl 中实现的逻辑有一些问题，有可能是访问了有冲突的资源、内存泄露等等。你可以先尝试把 XControl 中的不同部分的代码禁用一下，看看能不嗯能找到有问题的部分进行修改。

**Reply by tetsuoyyx** [reply]
好的，我试一下。还有个小问题请教一下，图 .21那个例子里面，为什么"Data Changed?"没有也更改为True？

**Reply by ruanqizhen** [reply]
最基本的原则是：这个 VI 上有两组输入输出控件“Data”和“Display State”，如果“Data”改变了，就应该把"Data Changed?"设置为 True；如果“Display State”改变了，就应该把"State Changed?"设置为 True。

**Reply by tetsuoyyx** [reply]
我明白您的意思，图 .21里面“Data”的输入到输出确实没有发生改变，但“Data”的输入和Xcontrol控件之前的数据（棋盘布局）不一样，chesses控件也被更新了，这个不是在改变数>据么，是我理解错了还是说这里面的改变就是指外观VI内部输入输出中间发生的改变？

**Reply by ruanqizhen** [reply]
这里的 Data 输入一定就是当前 Xcontrol 控件的数据。用户可能在界面上修改了控件的数据，但是这个事件分支输入的数据就已经是用户修改过之后的了。这个事件分支只要考虑，当前得到的数据是否需要再被修改。至于之前的改动，并不需要考虑。

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：


    XControl 是 LabVIEW 8 中出现的新功能。关于 XControl 功能介绍和实现方法可以参见这个网址：[http://zone.ni.com/devzone/conceptd.nsf/webmain/1AE46AF02D67AF468625706E006E577C](http://zone.ni.com/devzone/conceptd.nsf/webmain/1AE46AF02D67AF468625706E006E577C)。  
    面向组件的编程（Component Oriented Programming ，COP）技术建立在对象技术之上，它是对象技术的进一步发展。“类”这个概念仍然是组件技术中一个基础的概念，但是组件技术更核心的概念是“接口”。组件技术的主要目标是复用—粗粒度的复用，组件的核心是接口。

    LabVIEW 为我们提供了大量漂亮的控件，可以让我们非常方便地就搭建出一个程序界面来。然而，对于追求完美的用户而言，LabVIEW 提供的为数有限的控件是远远不够的。比如图1，是 LabVIEW 8.2 的一个新功能—导入导入共享库向导的界面。在它右上方有四个按钮，这四个按钮有着特殊的外观图标，在 LabVIEW 中并没有直接提供这样的按钮。要拥有这样的按钮，并保存下来以供再次使用，就只能自己来制作这样一个自定义控件。关于（用户自定义控件可以参考文章《[用户自定义控件中 Control, Type Def. 和 Strict Type Def. 的区别](http://ruanqizhen.spaces.live.com/blog/cns!5852D4F797C53FB6!1549.entry)》）

  ![](http://tkfiles.storage.msn.com/x1pN1mp8dKYgTFV_lNTUY6FnblHHsQ5D-NvZJN4vrpE0vnYl8YEKhrCyulbbcujRT7Ifp4UGJ50Rn1T1on4B2EKXM9n0bZkSSRXWb03q5IF6X7VdOI2u38s1A)  
图1：LabVIEW 8.2 中 Import Shared Library 的界面

    自定义控件虽然可以定义控件的外观，但无法定义控件的行为。功能复杂一点的控件，.ctl 文件就爱莫能助了。还是以图1 为例，它的 Include Paths 控件是“一个”功能比较复杂的控件，它比 LabVIEW 自带的列表框多了编辑功能。用户添加或编辑一个路径时，这个控件要为用户在所编辑的路径项目上，显示出供编辑使用的文本框和浏览路径按钮。  
    实际上这个编辑功能是由三个 LabVIEW 提供的标准控件合作完成的：一个 Listbox、一个 String 和一个 Button 控件。它们的行为（位置、是否可见，值，等等）是在程序运行时决定的：当用户选择编辑控件中某一路径时，程序就把 String 和 Button 挪到 Listbox 上需要编辑的那一项，并遮挡住 Listbox 原本的内容。这样，用户只能在 String 控件内输入内容，或者点击浏览按钮选择一个路径。编辑完成，程序把 String 控件的值写到 Listbox 上相应的项目中。

    我们虽然看不见图1 例子中的程序框图，但是可以想象，上述的一系列操作，如判断 String 和 Button 应当显示的位置；然后挪动它们、把 String 值传给 Listbox；处理用户对他们操作的消息等等，都会为这个程序添加不少复杂的代码。这些代码应该是与程序的其它部分没有任何直接关系的。但是如果把它们也写在这个界面 VI 的程序框图上，一方面影响了程序的可读性；另一方面，编程人员有可能在更改程序其它问题时，一不小心改变了这部分代码，降低了代码的安全性。  
    从逻辑关系上来看，图1 中上半部分的 Listbox、String、浏览按钮以及右上方四个操作按钮，合作共同完成一个功能，与它们之外的界面控件没有什么直接关联，所以他们七个应当被作为一个整体，或者说是一个组件。这个组件需要与程序其它模块之间的接口就只是一个字符串数组—用于输入或输出一组路径。其它的数据和操作，都应当是组件私有的、外部不可见的。  
    在 LabVIEW 8 之前，想分离和封装出这样一个组件是非常困难的。因为既然这七个控件都在这个 VI 的面板上，对它们的操作和相应的代码必须放在这个 VI 的程序框图上，无法与其他代码隔离开。当然也不是说绝对没有办法，比如你可以使用 sub-panel、动态注册事件等方法，强行地把它们的代码分隔开。但是这些方法既不简单直观，使用它们又可能会让程序变得更为复杂、难以阅读和维护。

    XControl 的出现终于为这个问题提供了一个比较完美的解决方案。图1 中我们提到的七个应当划分在同一组件的控件可以被制作成一个 XControl。这个 XControl 的外观就是图1 中上半部分七个控件组合在一起的样子。XControl 与用户自定义控件相比，它不仅定义了控件的外观，更重要的是，开发人员可以通过编写 LabVIEW 代码定义 XControl 的行为，这些代码是对外隐藏的。开发人员还可以定义 XControl 的属性和方法，通过 Property Node 和 Invoke Node 在程序中使用这些属性和方法。  
    同样完成选取一组路径这个功能，可以有各种不同的界面。比如各种 C++ 编译器都会提供类似的功能，但它们外观各不相同。你可以利用 XControl，编写多个外观、行为大相径庭的组件。但是，只要他们的接口相同—都是一个字符串数组，用户就可以在这些组件内任意互换，选用自己喜欢的组件，而不需改动程序的任何其它部分！  
    现在，XControl 仍然不太令人满意的地方是它还不支持用户自定义的事件。

    XControl 具有封装的特性。因此我在《[利用 LabVIEW 工程库实现面向对象编程](http://ruanqizhen.spaces.live.com/blog/cns!5852D4F797C53FB6!783.entry)》一文中提到，同样可以使用 XControl 来达到面向对象的编程方法。但是 XControl 不具备继承和多态的特性。与之相比较，Library 和 LVClass 只能够把程序中的某些功能封装成模块，而涉及到包含界面的模块，就无能为力了。XControl 则非常适合制作有界面的程序组件。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2009年07月6日 15:33 编辑
XControl有大量额外代码，效率肯定会受影响。2ms应当足以完成采集、显示工作了。但是这样做意义不大，为了提高效率应当每200ms采集100数据，而不是每2ms采集一个数据。

回复
gufei
2009年07月5日 21:38 编辑
阮老师，请问使用XControl会不会影响效率，我发现把代码直接写到vi中，比写到XControl中效率高很多啊、、、还有一个问题，我现在有个大数组，想模拟实时采集、处理显示，采样频率是500hz，即每2ms读取数组的一列，我的做法是用个循环，每2ms循环一次，处理和显示放在循环中，发现2ms内可能会处理不完操作，那就可能放弃操作了，请问有什么办法实现我的要求？

回复
gufei
2009年01月27日 10:59 编辑
提个建议啊，能不能给你的例子一些详细的解释：）感觉例子就没有什么解释啊，谢谢

回复
刘伟印
2008年09月5日 11:52 编辑
不错，支持一下

回复
Qizhen
2007年08月31日 13:29 编辑
最近我放这些程序的 google page 被政府和谐了。等我有时间换个地方放这些示例。

回复
Unknown
2007年08月31日 11:08 编辑
范例程序没有一个能下载的！！

回复
Zhang Ping
2007年07月25日 10:17 编辑
仰望一下 o(∩_∩)o…

回复
Unknown
2007年03月8日 13:55 编辑
受教了,如果多个XControl实例同时运行能准确无误就太好了.

回复
Wu
2006年11月23日 22:16 编辑
高！

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2009年09月16日 11:16 编辑
目前的控件已经满足用户普遍需求了。对于客户特殊化的需求，最好是由用户自定义完成。关于Picture控件，我觉得画点和线已经是最基本的操作了。公司内部的人使用这个控件，也就是调用它提供出来的这几个VI。Picture内部的实现，可能需要调用系统API提供的画图函数，但具体怎么实现的我也不是太了解。使用它的时候，其实也没必要了解它的内部格式。我从它的机构上猜测，Picture控件实际上记录的是用户的画图过程，它有自己的记录格式。基本格式是flatten后的簇数据。然后再内部调用系统的绘图函数，把图形绘制出来。当然，我没研究过这部分内容，只是自己猜测一下。

回复
彬
2009年09月15日 21:51 编辑
阮大侠： 对于labVIEW常用控件库从labVIEW7.0到LabVIEW2009都似乎没什么变化，而通过我的观察，控件库中添加最多的可以说是应用Picture控件自定义Xcontrol的3D控件了，由此趋势也可以发现Picture控件其实是最好用来自定义特殊界面的控件了，虽然labVIEW中开放了诸如：画点，画线等基本函数用来操作Picture控件画图，但还是不如直接深入工具包的内核探究其中如何操作Picture控件的。通过底层的代码我大概了解到Picture的一些特点，但只能管中窥豹，觉得学的很辛苦，请问是否能提供一些资料（Picture中的定义，格式）？

回复

WL
2008年07月16日 09:02 编辑
哎不懂

回复

死了都要C
2008年09月13日 10:38 编辑
阮大侠：
小弟新用LV，目前遇到一些难题，希望阮大侠能帮组解决。
1.程序运行速度的问题。目前做一个电路板的测试，测试项有很多，40多个大项，大项中有30多个小项，
程序的运行是采用动态调用实现的，刚开机运行速度还可以，但过一会非常慢，
是否与动态调用的资源释放有关系？
2.6608板卡测量两个信号跳变的滞后时间，
基本原理是用LV例程中的TWO EDGE SEPERATION MEASURENT.VI来实现的，硬件上采用先到信号连至AUX端，后发信号连至GATE端。
测量两个信号上升沿跳变的滞后时间是可行的，但测下降沿时始终是0，实在不知原因为何。
3.6608测量两个信号跳变的滞后时间的时机问题。
由于要测的这两个脚（PIN100.PIN101）信号的跳变是因另外一脚(PIN99)加高电平引起的，而这两个信号的跳变滞后时间非常短1MS左右，99脚加电平也是通过一数字板卡实现的（PXI6120），目前采取的方案是测量时间的VI跟99脚加电平并行执行，即将这两个VI的ERROR IN 并联。
请教阮大侠有没有更好的方案。
不胜感激。

回复

WL
2008年08月4日 09:56 编辑
专业啊

回复

Boone
2008年08月23日 22:26 编辑
我代表LabVIEW_Home，欢迎你来指导！
QQ群：13267790
Boone

回复
Qizhen
2008年08月5日 18:07 编辑
我后来没有吃任何药。只是加强锻炼，注意休息。症状缓解多了，现在只有偶尔才发病。

回复
文治
2008年08月5日 13:17 编辑
看了你的博客,我也有划痕症,请问你的好了没啊,怎么治的啊?
能加我QQ22623658么?或者邮件:liwenzhi84@163.com
不胜感谢

回复
